x-pack/plugins/alerts/server/authorization/alerts_authorization.ts

```diff
@@ -0,0 +1,281 @@
+      const checkPrivileges = this.authorization.checkPrivilegesDynamicallyWithRequest(
+        this.request
+      );
+
+      // add an empty `authorizedConsumers` array on each alertType
+      const alertTypesWithAutherization = this.augmentWithAuthorizedConsumers(alertTypes);
+
+      // map from privilege to alertType which we can refer back to when analyzing the result
+      // of checkPrivileges
+      const privilegeToAlertType = new Map<string, [RegistryAlertTypeWithAuth, string]>();
+      // as we can't ask ES for the user's individual privileges we need to ask for each feature
+      // and alertType in the system whether this user has this privilege
+      for (const alertType of alertTypesWithAutherization) {
+        for (const feature of featuresIds) {
+          privilegeToAlertType.set(
+            this.authorization!.actions.alerting.get(alertType.id, feature, operation),
+            [alertType, feature]
+          );
+        }
+      }
+
+      const { username, hasAllRequested, privileges } = await checkPrivileges([
+        ...privilegeToAlertType.keys(),
+      ]);
+
+      return {
+        username,
+        hasAllRequested,
+        authorizedAlertTypes: hasAllRequested
+          ? // has access to all features
+            this.augmentWithAuthorizedConsumers(alertTypes, [ALERTS_FEATURE_ID, ...featuresIds])
+          : // only has some of the required privileges
+            privileges.reduce((authorizedAlertTypes, { authorized, privilege }) => {
+              if (authorized && privilegeToAlertType.has(privilege)) {
+                const [alertType, consumer] = privilegeToAlertType.get(privilege)!;
+                alertType.authorizedConsumers.push(consumer);
+                authorizedAlertTypes.add(alertType);
+              }
+              return authorizedAlertTypes;
+            }, new Set<RegistryAlertTypeWithAuth>()),
+      };
+    }
+  }
+
+  private augmentWithAuthorizedConsumers(
+    alertTypes: Set<RegistryAlertType>,
+    authorizedConsumers?: string[]
+  ): Set<RegistryAlertTypeWithAuth> {
+    return new Set(
+      Array.from(alertTypes).map((alertType) => ({
+        ...alertType,
+        authorizedConsumers: authorizedConsumers ?? [ALERTS_FEATURE_ID],
+      }))
+    );
+  }
+
+  private asFiltersByAlertTypeAndConsumer(alertTypes: Set<RegistryAlertTypeWithAuth>): string[] {
+    return Array.from(alertTypes).reduce<string[]>((filters, { id, authorizedConsumers }) => {
+      ensureFieldIsSafeForQuery('alertTypeId', id);
+      filters.push(
+        `(alert.attributes.alertTypeId:${id} and (${authorizedConsumers
+          .map((consumer) => {
+            ensureFieldIsSafeForQuery('alertTypeId', id);
+            return `alert.attributes.consumer:${consumer}`;
+          })
+          .join(' or ')}))`
+      );
+      return filters;
+    }, []);
+  }
+}
+
+export function ensureFieldIsSafeForQuery(field: string, value: string): boolean {
+  const invalid = value.match(/[>=<\*:]+/g);
```

Comment:
Is there a reason to allow whitespace as part of the id or consumer value? If we disallow whitespace, then that will prevent something like `alert.attributes.alertTypeId: (foo or bar or baz)` from working. Might also be good to disallow `(` and `)` for the same reason

Reply:
that's a fair point, I just disallowed operators. I'll add those.